cmd/guacingest/cmd/ingest.go

```diff
@@ -80,69 +79,65 @@ func ingest(cmd *cobra.Command, args []string) {
 	}
 	defer csubClient.Close()
 
-	assemblerFunc, err := getAssembler(ctx, opts)
-	if err != nil {
-		logger.Errorf("error: %v", err)
-		os.Exit(1)
-	}
-
-	processorTransportFunc := func(d processor.DocumentTree) error {
-		docTreeBytes, err := json.Marshal(d)
-		if err != nil {
-			return fmt.Errorf("failed marshal of document: %w", err)
-		}
-		err = emitter.Publish(ctx, emitter.SubjectNameDocProcessed, docTreeBytes)
-		if err != nil {
-			return err
-		}
-		return nil
-	}
-
-	ingestorTransportFunc := func(d []assembler.IngestPredicates, i []*parser_common.IdentifierStrings) error {
-		err := assemblerFunc(d)
-		if err != nil {
-			return err
-		}
-
-		entries := input.IdentifierStringsSliceToCollectEntries(i)
-		if len(entries) > 0 {
-			logger.Infof("got collect entries to add: %v", len(entries))
-			if err := csubClient.AddCollectEntries(ctx, entries); err != nil {
-				logger.Errorf("unable to add collect entries: %v", err)
+	emit := func(d *processor.Document) error {
+		if opts.async {
+			docTree, err := process.Process(ctx, d)
+			if err != nil {
+				logger.Error("[processor] failed process document: %v", err)
+				return nil
 			}
-		}
-		return nil
-	}
 
-	processorFunc, err := getProcessor(ctx, processorTransportFunc)
-	if err != nil {
-		logger.Errorf("error: %v", err)
-		os.Exit(1)
-	}
+			docTreeBytes, err := json.Marshal(docTree)
+			if err != nil {
+				return fmt.Errorf("failed marshal of document: %w", err)
+			}
+			err = emitter.Publish(ctx, emitter.SubjectNameDocProcessed, docTreeBytes)
+			if err != nil {
+				logger.Error("[processor] failed transportFunc: %v", err)
+				return nil
+			}
 
-	ingestorFunc, err := getIngestor(ctx, ingestorTransportFunc)
-	if err != nil {
-		logger.Errorf("error: %v", err)
-		os.Exit(1)
+			logger.Infof("[processor] docTree Processed: %+v", docTree.Document.SourceInformation)
+			return nil
+		} else {
+			return ingestor.Ingest(ctx, d, opts.graphqlEndpoint, csubClient)
+		}
 	}
 
 	// Assuming that publisher and consumer are different processes.
 	var wg sync.WaitGroup
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		if err := processorFunc(); err != nil {
+		if err := process.Subscribe(ctx, emit); err != nil {
```

Comment:
in this case, even if it's not async, you will still need NATS running for this to work?

Reply:
meaning if `return ingestor.Ingest(ctx, d, opts.graphqlEndpoint, csubClient)` is running, why are we throwing it back into NATS here?

Reply:
Yes, this is the ingestor the receives documents via NATS. The only thing that async flag is doing is preventing it from sending it back via pub/sub via stages. 

So, sync scenario:

`doc -> nats -> ingestor.Ingest`

async scenario:

`doc -> nats -> processor.Process -> nats -> parser.Parse -> assembler`

I'm not sure if async scenario is having much sense any more, but I didn't want to remove it without further discussion.

Reply:
yea i am in favor of removing async for now unless we run into the need to granularly scale the processor and parser aspects, which i doubt we would need to in the near future.